ic-agent/src/agent/mod.rs

```diff
@@ -446,14 +459,80 @@ impl Agent {
-        let _envelope: Envelope =
+        let envelope: Envelope =
             serde_cbor::from_slice(&signed_query).map_err(AgentError::InvalidCborData)?;
-        self.query_endpoint::<QueryResponse>(effective_canister_id, signed_query)
-            .await
-            .and_then(|response| match response {
-                QueryResponse::Replied { reply } => Ok(reply.arg),
-                QueryResponse::Rejected(response) => Err(AgentError::ReplicaError(response)),
-            })
+        self.query_inner(
+            effective_canister_id,
+            signed_query,
+            envelope.content.to_request_id(),
+        )
+        .await
+    }
+
+    /// Helper function for performing both the query call and possibly a read_state to check the subnet node keys.
+    ///
+    /// This should be used instead of `query_endpoint`. No validation is performed on `signed_query`.
+    async fn query_inner(
+        &self,
+        effective_canister_id: Principal,
+        signed_query: Vec<u8>,
+        request_id: RequestId,
+    ) -> Result<Vec<u8>, AgentError> {
+        let response = if self.verify_query_signatures {
+            let (response, subnet) = futures_util::try_join!(
+                self.query_endpoint::<QueryResponse>(effective_canister_id, signed_query),
+                self.get_subnet_by_canister(&effective_canister_id)
+            )?;
+            if response.signatures().is_empty() {
+                return Err(AgentError::MissingSignature);
+            }
+            for signature in response.signatures() {
+                if OffsetDateTime::now_utc()
+                    - OffsetDateTime::from_unix_timestamp_nanos(signature.timestamp as _).unwrap()
+                    > self.ingress_expiry
+                {
+                    return Err(AgentError::CertificateOutdated(self.ingress_expiry));
+                }
+                let signable = response.signable(request_id, signature.timestamp);
+                let node_key = subnet
+                    .node_keys
+                    .get(&signature.identity)
+                    .ok_or(AgentError::CertificateNotAuthorized())?;
+                if node_key.len() != 44 {
+                    return Err(AgentError::DerKeyLengthMismatch {
+                        expected: 44,
+                        actual: node_key.len(),
+                    });
+                }
+                const DER_PREFIX: [u8; 12] = [48, 42, 48, 5, 6, 3, 43, 101, 112, 3, 33, 0];
+                if node_key[..12] != DER_PREFIX {
+                    return Err(AgentError::DerPrefixMismatch {
+                        expected: DER_PREFIX.to_vec(),
+                        actual: node_key[..12].to_vec(),
+                    });
+                }
+                let pubkey =
+                    VerificationKey::try_from(<[u8; 32]>::try_from(&node_key[12..]).unwrap())
+                        .map_err(|_| AgentError::QuerySignatureVerificationFailed)?;
```

Comment:
Suggestion: use distinct errors (or differentiate by an enum member or something) rather than three possibilities that all end up as `QuerySignatureVerificationFailed`.  It's frustrating enough to investigate a signature verification failure when the only possibility is that the verify() method failed
```suggestion
                        .map_err(|_| AgentError::MalformedVerificationKey)?;
```